### DISCUSSION TITLE
Add BlurHash for Images

### DISCUSSION THREAD

**Discussion by madc** [directus/directus#8237] [Feature Requests]
It would be nice to have BlurHash (see [blurha.sh](https://blurha.sh/)) generated for images.
Generation could happen on upload.

I have not used it personally before, but think it could be a nice addition for statically generated websites.
What do you think?

**Comment by joselcvarela** [comment]
Wow, this is a very nice technique.

@rijkvanzanten @benhaynes what are your thoughts? Basically this can generate a string < 30 chars and show a blur placeholder (like a base64 image) while image is loading.

**Reply by rijkvanzanten** [reply]
Hmm it's a nice bit of tech, but it's very proprietary (eg this hash string isn't any standard). You can already generate blurred images from Directus' exiting image transformation capabilities (you could use /assets in combination with width:30 height:30 and blur:5 for example to create a similar placeholder image for use in your static website f.e. 🤔 

I'm wondering if this is more the realm of a custom hook than a core feature (80/20)

**Reply by joselcvarela** [reply]
Yeah I have checked `sharp` package and found out the blur property.
But with a string it would be much more economic on computing resources.

I need to agree that is too much specific, even though also argon2 is 😛

**Comment by madc** [comment]
Thanks for the feedback. I reviewed our project with it in mind and tend to agree with @rijkvanzanten.
We use a custom implementation for Nuxt to generate static images. While I don't think, the blur filter is a good replacement for BlurHash (thanks for pointing it out tough), the generation of the hash could (should?) happen locally at build time.

**Reply by rijkvanzanten** [reply]
> the generation of the hash could (should?) happen locally at build time.

Yeah!

Another thing that I just thought of is that you could do it on the directus side of things in a custom hook. You could add a new field to directus_files to hold the hash, and then setup a "files.upload" hook to auto-generate the blurhash the moment the file is uploaded 🙂

**Reply by madc** [reply]
Oh, that's just beautiful. I had a quick shot at it and got quite far. I can read the image and create the hash on upload.
Only thing is, can you point me in the direction how to update the file record in a hook and save it?

**Reply by rijkvanzanten** [reply]
It should be fairly similar to this example: https://docs.directus.io/guides/api-endpoints/#full-example but using the files service.

Try—and I didn't test this exact snippet, so your mileage may vary—something like:

```js
module.exports = registerHooks({ services }) {
  const { FilesService } = services;

  return {
    'files.upload': async function({ item, schema, payload }) {
      const service = new FilesService({ schema });

      // this is what you already have i think
      const blurHashResult = await getBlurHash(payload);
     
      await service.updateOne(item, {
        blur_hash: blurHashResult
      });
    }
  }
}
```

**Reply by madc** [reply]
Works exactly as it is. Thank you for your support, this is awesome!

**Comment by madc** [comment]
This is a quick, but working  integrationof BlurHash as `files.upload` hook.

**Requirements**:
- [NodeJS BlurHash module](https://github.com/woltapp/blurhash) is manually added to the project
- A field named `blurhash` (String, 30) was added to the system collection `directus_files`.

`extensions/hooks/blurhash/index.js`
```js
const sharp = require('sharp');
const { encode } = require('blurhash');
const { join } = require('path');

module.exports = function registerHook({ env, services }) {
    return {
        'files.upload': async function ({ item, schema, payload }) {
            if (['image/jpeg', 'image/png', 'image/webp', 'image/gif', 'image/tiff'].includes(payload.type)) {
                // const buffer = await storage.disk(payload.storage).getBuffer(payload.filename_disk);
                const buffer = await sharp(join(env.STORAGE_LOCAL_ROOT, payload.filename_disk))
                    .raw()
                    .ensureAlpha()
                    .toBuffer();

                const blurhash = encode(new Uint8ClampedArray(buffer), payload.width, payload.height, 4, 4)

                const { FilesService } = services;
                const service = new FilesService({ schema });
                await service.updateOne(item, { blurhash });
            }
        },
    };
};
```

**Caveats**:
- Currently it does not use the internal storage to load the buffer, because I could not access it inside the hook. This means, it only works for `local` files.
- The original image (`filename_disk`) is used to create the BlurHash, which can be slow. This could be improved by using a thumbnail.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
